hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1078,30 +1111,49 @@ private class Emit(
     }
   }
 
-  private def getAsDependentFunction[A1: TypeInfo, R: TypeInfo](
-    ir: IR, argname: String, env: Emit.E, fb: EmitFunctionBuilder[_], errorMsg: String
-  ): DependentFunction[AsmFunction3[Region, A1, Boolean, R]] = {
+  private def capturedReferences(ir: IR): (IR, (Emit.E, DependentEmitFunction[_]) => Emit.E) = {
     var ids = Set[String]()
 
     def getReferenced: IR => IR = {
-      case Ref(id, typ) if id == argname =>
-        In(0, typ)
-      case node@Ref(id, _) if env.lookupOption(id).isDefined =>
+      case node@Ref(id, typ) =>
         ids += id
         node
       case node => MapIR(getReferenced)(node)
     }
```

Comment:
hmm, good point! The binding shadowing is not handled correctly here, but I don't think this can cause incorrect behavior anywhere. The worst-case scenario is that a few extra values get copied into the dependent function class fields because the bindings get shadowed by something else inside the dependent function, and the shadowing is still handled correctly inside `Emit`.

e.g. in 

```
Let("foo", I32(0),
    Uniroot("x", 
        Let("foo", F64(5.5), Ref("foo", TInt64())), 
        F64(0.0), F64(1.0))
```
the function would extract "foo" as a referenced binding, even though it (the integer foo) is never used in the function, and store the value in the class. When we actually emit the function, though, `Emit` handles the binding shadowing correctly and uses the correct `foo` (the one defined in the second Let) to actually calculate the value of the function.

I'm inclined to leave it as-is because copying extra field values is cheap, and I don't think there is a super easy way to take that into account at this level with what we have.

Reply:
(There's a test with internal `Lets` in the Uniroot tests. I can add one where two bindings have the same name, if you'd like, to test this.